File: core/scala/src/main/scala/org/aiddl/core/scala/function/misc/LambdaFunctionEvaluator.scala
package org.aiddl.core.scala.function.misc

import org.aiddl.core.scala.container.Container
import org.aiddl.core.scala.eval.Evaluator
import org.aiddl.core.scala.function.{Function, LazyFunction, DefaultFunctionUri as D}
import org.aiddl.core.scala.representation.*

object LambdaFunctionEvaluator {
  private var NextID = 0
}

protected[function] class LambdaFunctionEvaluator(c: Container) extends Function with LazyFunction {
  private class LambdaFunction(x: Term, f: Term, eval: Evaluator) extends Function {

    override def apply(arg: Term): Term = x unify arg match {
      case Some(s) => eval(f \ s)
      case None => throw new IllegalArgumentException(arg.toString + " does not match argument pattern of " + this.toString())
    }

    override def toString(): String = "(org.aiddl.eval.lambda " + this.x + " " + this.f + ")"
  }

  override def apply(x: Term): Term = x match {
    case Tuple(arg, fun) => {
      val f = new LambdaFunction(arg, fun, c.eval)
      val uri = Sym("#lambda_" + LambdaFunctionEvaluator.NextID)
      LambdaFunctionEvaluator.NextID += 1
      c.addFunction(uri, f)
      FunRef(uri, f)
    }
    case _ => throw new IllegalArgumentException(s"Bad argument: $x. Expected tuple (arg fun).")
  }
}

File: core/java/src/main/java/org/aiddl/core/java/function/misc/LambdaFunctionEvaluator.java
package org.aiddl.core.java.function.misc;


import org.aiddl.core.java.eval.Evaluator;
import org.aiddl.core.java.function.FunctionRegistry;
import org.aiddl.core.java.function.Lambda;
import org.aiddl.core.java.function.Uri;
import org.aiddl.core.java.interfaces.Function;
import org.aiddl.core.java.interfaces.LazyFunction;
import org.aiddl.core.java.representation.SymbolicTerm;
import org.aiddl.core.java.representation.Term;

public class LambdaFunctionEvaluator implements LazyFunction {
	
	static int NextID = 0;
	
	FunctionRegistry fReg;
	Evaluator eval;
	
	public LambdaFunctionEvaluator( FunctionRegistry fReg ) {
		this.fReg = fReg;
		this.eval = (Evaluator)fReg.getFunction(Uri.EVAL);
	}

	@Override
	public Term apply(Term args) {
		Term arg_term = args.get(0);
		Term fun_term = args.get(1);

		Function f = new Lambda(arg_term, fun_term, this.eval);
		SymbolicTerm uri = Term.sym("#lambda_" + NextID);
		NextID += 1;
		this.fReg.addFunction(uri, f);
		
		return Term.fref(uri, fReg);
	}
}
